Support arithmetic operations on numerical fields for scalar filtering

avsolatorio opened the pull request:
This enhancement will allow users to include binary arithmetic operations on fields defined in filter expressions passed into the predicate of a query.

The following examples of expressions are now supported after this patch.

```
expr="(int64 + 10 == 29)"
expr="(int64 - 64 == 2016)"
expr="(int64 * 59 == 1029)"
expr="(int64 / 11 == 2)"
expr="(int64 % 11 == 0)"
expr="(int64 / 11 == 2) || int64 in [1, 2, 3, 4]"
```

Allowing arithmetic operations on fields can allow us to implement complex heuristics based on the encoding of the fields.  For example, if we encode multi-valued fields as products of prime numbers, we can effectively filter these fields using the modulo of the corresponding prime number assigned to the value in the multi-valued field.

Concretely, say we have the data `data1=["Philippines", "Japan", "China", "USA"]` and `data2=["Indonesia", "Malawi", "China", "Philippines"]` which we represent as prime numbers:

```
Philippines -> 2
Japan -> 3
China -> 5
USA -> 7
Indonesia -> 11
Malawi -> 13
```

we encode these data as product of primes `p_data1=210` and `p_data2=2002` and store them in an INT64 field `p_countries`.

Now, if a user wants to find data that contain "Philippines", we can simply take the prime associated with the input (Philippines=2) and perform the expression `(p_countries % 2 == 0)`. This will be true for both `p_data1` and `p_data2`.

Alternatively, if a user wants to find the data that contain "Malawi", we use the prime (Malawi=13) and perform the expression `(p_countries % 13 == 0)`. This will return only `p_data2`.

This method will generally permit users to store multi-valued data, albeit ones with low cardinality (due to the INT64 limit), and filter them within Milvus (no need to fetch batches and filter post query). This is helpful for data with labels that are rare, which are likely to be missed out if we don't filter directly the data.

Signed-off-by: Aivin V. Solatorio <[avsolatorio@gmail.com](mailto:avsolatorio@gmail.com)>
/kind feature
issue: https://github.com/milvus-io/milvus/issues/16513

avsolatorio committed:
Support arithmetic operations on numerical fields for scalar filtering

Signed-off-by: Aivin V. Solatorio <avsolatorio@gmail.com>

xiaofan-luan commented:
@DragonDriver please help on review this issue?

xiaofan-luan commented:
hi @avsolatorio , please take a look a look at the go format issue and code coverage~ thanks my man

avsolatorio commented:
Hello @xiaofan-luan, I have added new test cases to hopefully increase the coverage and I also fixed the go format issue. I believe the code checks and codecov report require manual approval to rerun, right? Thanks!

DragonDriver commented on internal/core/src/index/ScalarIndex.h:
```diff
@@ -31,6 +31,12 @@ class ScalarIndex : public IndexBase {
     NotIn(size_t n, const T* values) = 0;
 
+    virtual const TargetBitmapPtr
+    EvalEq(std::string arith_op, T right_operand, T value) = 0;
```
I don't think `EvalEq` and `EvalNotEq` should be defined in an index. Maybe we should use raw-data to do these two filterings. As you have already implemented, in fact, many index type can't support them.

avsolatorio replied:
RIght! I removed this in the updated version. I followed your advice of processing the data directly using the `elem_func` construct.

DragonDriver commented on internal/core/src/query/Expr.h:
```diff
@@ -112,6 +112,49 @@ enum class OpType {
 };
 
+enum class ArithOpType {
+    Unk = 0,
```
`Unk` here is a little confusing. Consider `Unknown` instead.

DragonDriver commented on internal/core/src/query/visitors/ExecExprVisitor.cpp:
```diff
@@ -222,6 +226,114 @@ ExecExprVisitor::ExecUnaryRangeVisitorDispatcher(UnaryRangeExpr& expr_raw) -> Bi
+        case OpType::Equal: {
+            switch (arith_op) {
+                case ArithOpType::Add: {
+                    auto index_func = [val, right_operand, arith_op](Index* index) {
```
As above mentioned, I suggest that we simply use raw data to do brute-force search here. You can abstract a newly function like `ExecRangeVisitorImpl`, this function accepts a `ElemFunc` and will apply the `ElemFunc` for every row in every chunk in the segment.

avsolatorio replied:
I implemented an `ExexDataRangeVisitorImpl` that does the filtering of the rows. It passed the tests, but I'd like to ask if the changes I made are what you had in mind, too. :)

DragonDriver replied:
Yes! Well-matched what I wanted. thx.

xiaofan-luan commented:
> Hello @xiaofan-luan, I have added new test cases to hopefully increase the coverage and I also fixed the go format issue. I believe the code checks and codecov report require manual approval to rerun, right? Thanks!

Sure, I will help on the rerun

xiaofan-luan commented:
/assign

DragonDriver commented on internal/core/unittest/test_expr.cpp:
```diff
@@ -581,3 +581,364 @@ TEST(Expr, TestCompare) {
     }
 }
+
+TEST(Expr, TestBinaryArithOpEvalRange) {
```
Please also add ut for expr. As for how to generate plan from expr conviniently, maybe you can refer to `plan_parser_test.go` and `translate_text_plan_to_binary_plan`.

DragonDriver replied:
Please also add ut for `SegmentInterface.Search` and `SegmentInterface.Retrieve`.
Also, an e2e test demo under `tests/python_client/testcases/` would be perfect.

DragonDriver commented on internal/proto/plan.proto:
```diff
@@ -14,6 +14,15 @@ enum OpType {
 };
 
+enum ArithOpType {
+  Unk = 0;
```
As mentioned above, `Unk` here is a little confusing.

DragonDriver commented on internal/proto/plan.proto:
```diff
@@ -82,6 +91,20 @@ message BinaryExpr {
   Expr right = 3;
 }
 
+message BinaryArithOp {
```
It seems that `BinaryArithOp` is not used. If so, please delete it.

DragonDriver commented on internal/proxy/plan_parser.go:
```diff
@@ -72,8 +72,17 @@ func (optimizer *optimizer) Exit(node *ant_ast.Node) {
+		// Check IdentifierNodes
+		identifierNodeLeft, leftIdentifier := node.Left.(*ant_ast.IdentifierNode)
+		identifierNodeRight, rightIdentifier := node.Right.(*ant_ast.IdentifierNode)
+		funcName, err := getFuncNameByNodeOp(node.Operator)
```
What if the `node.Operator` is not of arithmetic operator?

avsolatorio replied:
You're right, I have moved this inside each operation that needs it.

DragonDriver commented on internal/proxy/plan_parser.go:
```diff
@@ -517,6 +679,12 @@ func (pc *parserContext) handleMultiCmpExpr(node *ant_ast.BinaryNode) (*planpb.E
+	_, arithExpr := node.Left.(*ant_ast.FunctionNode)
+
+	if arithExpr {
+		return pc.handleBinaryArithCmpExpr(node)
```
It's ok that our optimizer handles the ExprTree well (optimizer transfer arithmetic-related node to FunctionNode). You can also consider `BinaryArithCmpExpr` as a special cmp expr and then handle it in cmp expr recursively.
Both these ways make sense for me.

avsolatorio replied:
Ah, I decided to implement the `FunctionNode` since I find it to aptly represent what the feature wanted to accomplish, i.e., to analyze the expression as a function. Nonetheless, I will think through how to better design this and potentially implement the `BinaryArithCmpExpr`  in the next PR if it serves to be a convenient choice when expanding the functionality to support other comparison operators. Also, I remember that I ended up doing the `FunctionNode` transform because, originally, the optimizer will raise an exception if an arithmetic operator has an `IdentifierNode` as its operands. So I had to handle it at the optimizer. 😄

DragonDriver commented on internal/core/src/query/visitors/ExecExprVisitor.cpp:
```diff
@@ -222,6 +226,114 @@ ExecExprVisitor::ExecUnaryRangeVisitorDispatcher(UnaryRangeExpr& expr_raw) -> Bi
+    auto val = expr.value_;
+
+    switch (op) {
+        case OpType::Equal: {
```
Besides `==` and `!=`, should we support `<`, `<=`, `>`, `>=` also?

avsolatorio replied:
Yes, I am thinking of adding these additional comparison operators in a separate PR. I will need to think more about how to better integrate these into the overall architecture of the parser. And potentially refactor the current "==" and "!=" handling, if needed. I hope this is fine. :)

DragonDriver commented:
@avsolatorio lgtm, also you can complete some details in later pr.

czs007 commented on internal/core/src/query/visitors/ExecExprVisitor.cpp:
```diff
@@ -174,6 +178,31 @@ ExecExprVisitor::ExecRangeVisitorImpl(FieldOffset field_offset, IndexFunc index_
+        for (int index = 0; index < this_size; ++index) {
+            result[index] = element_func(data[index]);
+        }
+        AssertInfo(result.size() == this_size, "");
```
Please provide more detailed exception assertion information.

avsolatorio replied:
Hello @czs007, I have just added a more verbose exception message for this. Thank you!

avsolatorio commented:
> @avsolatorio lgtm, also you can complete some details in later pr.

@DragonDriver, I'll surely create a separate PR for expanding the functionality and also adding more tests. Thank you! 😄

DragonDriver commented:
rerun ut

avsolatorio commented:
Hi @DragonDriver , thanks for rerunning the test! I see that it failed again. I'm not sure if the failing test is one of those that intermittently fail or if the failure was caused by my changes. wdyt? 🤔

DragonDriver commented:
@avsolatorio The ut failure is not caused by your changes. We're fixing it.

DragonDriver commented:
@avsolatorio The fix of ut has already been merged. You can try rebasing the latest branch to trigger the github action again.

avsolatorio commented:
@DragonDriver thanks for the heads up, it's passing now. 😄

avsolatorio commented:
Hello @DragonDriver , the workflows are still failing. Anything that I can do on my end? Also, once the build succeeds, will this PR be merged already? Thank you! cc: @xiaofan-luan

DragonDriver commented:
rerun ut

czs007 commented:
@avsolatorio good job!  Thanks!
/approve
/lgtm

avsolatorio commented:
Oh, cool! Thank you, everyone! 😄

xiaofan-luan commented:
> Oh, cool! Thank you, everyone! 😄

Thanks for the contribution my man, impressive patch~
